src/Workspaces/Core/Portable/SemanticModelReuse/SemanticModelWorkspaceServiceFactory.SemanticModelWorkspaceService.cs

```diff
@@ -0,0 +1,175 @@
+                // model value in it.
+                var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+                var builder = ImmutableDictionary.CreateBuilder<DocumentId, SemanticModelReuseInfo?>();
+
+                // Note: we are intentionally storing the semanticmodel instance in the speculative location as well.
+                builder.Add(document.Id, new SemanticModelReuseInfo(semanticModel, semanticModel, bodyNode, topLevelSemanticVersion));
+
+                foreach (var linkedId in linkedIds)
+                    builder.Add(linkedId, null);
+
+                return builder.ToImmutable();
+            }
+
+            private static async Task<SemanticModelReuseInfo?> TryReuseCachedSemanticModelAsync(
+                ImmutableDictionary<DocumentId, SemanticModelReuseInfo?> map,
+                Document document,
+                SyntaxNode bodyNode,
+                ImmutableArray<DocumentId> linkedIds,
+                VersionStamp topLevelSemanticVersion,
+                CancellationToken cancellationToken)
+            {
+                // Get the clique (i.e. all the ids for this and the other documents it is linked to) corresponds to
+                // matches the clique we're caching.
+                using var _ = PooledHashSet<DocumentId>.GetInstance(out var documentIdClique);
+                documentIdClique.Add(document.Id);
+                documentIdClique.AddRange(linkedIds);
+
+                // if this is asking about a different doc or clique, we can't reuse anything.
+                if (!map.Keys.SetEquals(documentIdClique))
+                    return null;
+
+                // see if this doc matches the docs we're caching information for.
+                if (!map.TryGetValue(document.Id, out var reuseInfoOpt) || !reuseInfoOpt.HasValue)
+                    return null;
+
+                var reuseInfo = reuseInfoOpt.Value;
+
+                // can only reuse the cache if nothing top level changed.
+                if (reuseInfo.TopLevelSementicVersion != topLevelSemanticVersion)
+                    return null;
+
+                // If multiple callers are asking for the exact same body, they can share the exact same semantic model.
+                // This is valuable when several clients (like completion providers) get called at the same time on the
+                // same method body edit.
+                if (reuseInfo.BodyNode == bodyNode)
+                    return reuseInfo;
+
+                var reuseService = document.GetRequiredLanguageService<ISemanticModelReuseLanguageService>();
+                var semanticModel = await reuseService.TryGetSpeculativeSemanticModelAsync(reuseInfo.PreviousNonSpeculativeSemanticModel, bodyNode, cancellationToken).ConfigureAwait(false);
+                if (semanticModel == null)
+                    return null;
```

Comment:
See earlier question again: if one side on the multi-targeting can't reuse, we're dropping both sides?

Reply:
fixed up to reuse the other side.